cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileMetadataReader.java

```diff
@@ -0,0 +1,225 @@
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = framework;
+    this.rootLocationFactory = rootLocationFactory;
+    this.impersonator = impersonator;
+    this.metaTableName = LoggingConfiguration.LOG_META_DATA_TABLE;
+  }
+
+
+  // This is only used in testing
+  public FileMetadataReader(final DatasetFramework framework,
+                            final TransactionSystemClient txClient,
+                            final RootLocationFactory rootLocationFactory,
+                            final Impersonator impersonator,
+                            final String metaTableName) {
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(framework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = framework;
+    this.rootLocationFactory = rootLocationFactory;
+    this.impersonator = impersonator;
+    this.metaTableName = metaTableName;
+  }
+
+  private Table getLogMetaTable(DatasetContext datasetContext) throws IOException, DatasetManagementException {
+    DatasetId metaTableInstanceId = NamespaceId.SYSTEM.dataset(metaTableName);
+    return DatasetsUtil.getOrCreateDataset(datasetContext,
+                                           datasetFramework, metaTableInstanceId, Table.class.getName(),
+                                           DatasetProperties.EMPTY);
+  }
+
+  private byte[] getRowKey(LogPathIdentifier logPathIdentifier) {
+    return Bytes.add(ROW_KEY_PREFIX, logPathIdentifier.getRowKey().getBytes());
+  }
+
+  /**
+   * Returns a list of log files for a logging context.
+   *
+   * @param logPathIdentifier logging context identifier.
+   * @param startTimestampMs starting timestamp in milli seconds
+   * @param endTimestampMs ending timestamp in milli seconds
+   * @return List of {@link LogLocation}
+   */
+  public List<LogLocation> listFiles(final LogPathIdentifier logPathIdentifier,
+                                     final long startTimestampMs, final long endTimestampMs) throws Exception {
+
+    return txExecute(transactional, new TxCallable<List<LogLocation>>() {
+      @Override
+      public List<LogLocation> call(DatasetContext context) throws Exception {
+        Table table = getLogMetaTable(context);
+        final Row cols = table.get(getRowKey(logPathIdentifier));
+
+        if (cols.isEmpty()) {
+          //noinspection unchecked
+          return new ArrayList<>();
+        }
+
+        final List<LogLocation> files = new ArrayList<>();
+
+        for (Map.Entry<byte[], byte[]> entry : cols.getColumns().entrySet()) {
+          // the location can be any location from on the filesystem for custom mapped namespaces
+          if (entry.getKey().length == 8) {
```

Comment:
It would be good to add more comments around `entry.getKey().length` length being 8 vs 17.